### PR TITLE
Allow Module forward-pre and forward hooks to take kwargs

### LINKED ISSUE #35643: nn.Module hooks ignore kwargs
To reproduce:

https://gist.github.com/zjplab/1d7b2a22a13d63f3ad83a59563cc4141

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are @ezyang @SsnL @zou3519 @gqchen

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89389

closes #35643

This PR is mostly borrowed from #82042. Thanks @Padarn for implementing
the first version and debugging into the errors.

Based on the discussion in #82042 this PR adds a with_kwargs
argument to register_forward_pre_hook and register_forward_hook
methods. When the arg is set to true, the provided hook must accept
kwargs args. Under the hook, this PR adds a
`_forward_pre_hooks_with_kwargs` and a `_forward_hook_with_kwargs`
set to keep track of which hooks accept kwargs.

Differential Revision: [D41431111](https://our.internmc.facebook.com/intern/diff/D41431111)

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### CODE REVIEW

**Comment by Padarn** [comment]
Thaks @mrshenli! I've closed my PR. I like the changes you've made 👍

File: torch/nn/modules/module.py
Comment (Padarn) [inline torch/nn/modules/module.py:89]: Yeah I had this same realisation. It does feel like they should be mergable, but I couldn't see a simple way to do so.

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

File: torch/nn/modules/module.py
Comment (awgu) [inline torch/nn/modules/module.py:134]: nit:
```suggestion
# N.B.: This class is NOT deriving from `_WrappedHook`, because pre- and post-
```

File: torch/nn/modules/module.py
Comment (awgu) [inline torch/nn/modules/module.py:1555]: nit:
```suggestion
            # N.B.: Wrapping OrderedDict values into a list is necessary as
```

File: torch/nn/modules/module.py
Comment (awgu) [inline torch/nn/modules/module.py:1565]: nit: 
```suggestion
            # N.B.: Wrapping OrderedDict values into a list is necessary as
```

**Review by awgu — COMMENTED** [review]
The approach makes sense to me, and this looks good to unblock `fully_shard()` :)

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

**Review by soulitzer — COMMENTED** [review]
The idea sounds good, but I think there are still two points where this is bc-breaking. (Since this is a new feature, not being bc-breaking at all is preferred.)
- If someone loads a module saved with a previous version of torch sounds like this would currently error? Could we automatically wrap old-style hooks in the new wrapper during forward?
- Is there a way to mitigate the use case where someone calls into _forward_hooks and checks isinstance? From the test failures it looks like a common use case.

File: torch/nn/modules/module.py
Comment (mrshenli) [reply torch/nn/modules/module.py:1423]: > Is there a way to mitigate the use case where someone calls into _forward_hooks and checks isinstance? From the test failures it looks like a common use case.

How about making `_forward_hooks` an `@property`, and unwrap the original hooks within the property method?

Comment (mrshenli) [reply torch/nn/modules/module.py:1423]: maybe the `@property` `_forward_hooks` method can return a custom type that overrides `__getitem__`?

**Comment by soumith** [comment]
please wait for @albanD 's review before merging this in any form

**Comment by albanD** [comment]
We discussed this one with @soulitzer and you can see the comment above.
In particular, all the modifications to the AO code are a good indication that 1) users poke at the implementation of these things pretty hard and 2) this implementation is actually BC-breaking in practice.
Given that this is a brand new feature, it should be fully backward compatible and the AO code shouldn't need to be modified in this PR.

In particular, an extra if-statement in the `__call__` method sounds like an ok tradeoff to be fully BC.

File: torch/nn/modules/module.py
Comment (soulitzer) [reply torch/nn/modules/module.py:1423]: Yes, though this custom type may also need to implement other methods like `.items()`, `.keys()`, `.values()`, `.get()`, etc don't automatically call into `__getitem__`.

If that gets too messy, perhaps storing the original hooks in _forward_hooks and maintaining a mapping of which ones accept kwargs directly on the module would be a better approach?

**Comment by mrshenli** [comment]
Got it. Let me see if I can keep BC for `_forward_pre_hooks` and `_forward_hooks`.

File: torch/nn/modules/module.py
Comment (mrshenli) [reply torch/nn/modules/module.py:1423]: Sure, let me try this

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

File: torch/nn/modules/module.py
Comment (mrshenli) [inline torch/nn/modules/module.py:424]: One caveat is that when users directly access `_forward_hooks` to delete hooks, the counterpart in `_forward_hooks_with_kwargs` won't be deleted accordingly. This should be fine as `_forward_hooks_with_kwargs` only applies to new hooks so no BC issue?

@albanD @soulitzer @awgu @Padarn

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

**Comment by mrshenli** [comment]
@albanD  @soulitzer  @awgu  @Padarn  this is ready for another round of review

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

File: torch/nn/modules/module.py
Comment (awgu) [inline torch/nn/modules/module.py:421]: nit: dangling "All"?
```suggestion
    # Marks whether the corresponding _forward_hooks accept kwargs or not.
```
(same with below for `_forward_pre_hooks_with_kwargs`)

File: torch/nn/modules/module.py
Comment (soulitzer) [reply torch/nn/modules/module.py:424]: That sounds OK. But we may still want the corresponding entry in `_forward_hooks_with_kwargs` to be deleted if someone properly uses the handle to remove the hook.

File: torch/nn/modules/module.py
Comment (soulitzer) [inline torch/nn/modules/module.py:1491]: nit: pre-hook

File: torch/nn/modules/module.py
Comment (soulitzer) [inline torch/nn/modules/module.py:1372]: "passed with" -> "passed the"

File: torch/nn/modules/module.py
Comment (soulitzer) [inline torch/nn/modules/module.py:1357]: Maybe change to "... given to the forward function. And if the hook modifies the input, both the args and kwargs should be returned."

**Review by soulitzer — COMMENTED** [review]
Thanks for the update, mostly looks good! Mainly just a comment about cleaning up the new dict on removal.

**Comment by mrshenli** [comment]
Thanks @awgu and @soulitzer for the review! Before updating this PR, will let CI run for a bit longer while waiting for comments from other reviewers

File: torch/nn/modules/module.py
Comment (albanD) [inline torch/nn/modules/module.py:1348]: nit: `False`

File: torch/nn/modules/module.py
Comment (albanD) [inline torch/nn/modules/module.py:1402]: nit: you could move this signature just at the end of the with_kwargs=False paragraph below. So that each paragraph has the signature for the hook next to it.

File: torch/nn/modules/module.py
Comment (albanD) [inline torch/nn/modules/module.py:1492]: nit: `(new_args, new_kwargs)`

File: torch/nn/modules/module.py
Comment (albanD) [reply torch/nn/modules/module.py:424]: Yes the handle.remove() method should remove the entry in the _forward_hooks_with_kwargs dict.

File: torch/utils/hooks.py
Comment (albanD) [inline torch/utils/hooks.py:16]: Not sure what `kwargs_dict` means in this context. `extra_dict` sounds better.
Also the typing here should be the same as `hooks_dict` no?

Even better might be to just do `*args` and take as many dictionary as the user wants.

File: torch/utils/hooks.py
Comment (albanD) [inline torch/utils/hooks.py:51]: Why dict() nad not OrderedDict() like the one above?

File: torch/utils/hooks.py
Comment (mrshenli) [reply torch/utils/hooks.py:16]: Let me do `extra_dict` for now. Can add arbitrary numbers of dict if necessary in followups.

File: torch/utils/hooks.py
Comment (mrshenli) [reply torch/utils/hooks.py:51]: I thought we don't need order for this one. Let me change that to `OrderedDict` to keep things consistent.

Comment (albanD) [reply torch/utils/hooks.py]: Same as with the other one tbh.

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

File: torch/nn/modules/module.py
Comment (albanD) [inline torch/nn/modules/module.py:1498]: nit: raise RuntimeError not plain asserts

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

**Review by albanD — COMMENTED** [review]
Small nit about error but SGTM otherwise.
Will let @soulitzer take a final look as well!

**Comment by mrshenli** [comment]
@mrshenli has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41431111).

**Review by soulitzer — APPROVED** [review]
Looks good, thanks!

**Comment by mrshenli** [comment]
Regarding the errors:

1. `
periodic / linux-bionic-cuda11.6-py3-gcc7-slow-gradcheck` https://github.com/pytorch/pytorch/actions/runs/3527608027/jobs/5917112620

```
2022-11-22T23:52:48.9571582Z ======================================================================
2022-11-22T23:52:48.9571873Z ERROR [0.032s]: test_index_select_exhaustive_index_large_cuda_float64 (__main__.TestSparseCUDA)
2022-11-22T23:52:48.9572149Z ----------------------------------------------------------------------
2022-11-22T23:52:48.9572285Z Traceback (most recent call last):
2022-11-22T23:52:48.9572640Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2054, in wrapper
2022-11-22T23:52:48.9572757Z     method(*args, **kwargs)
2022-11-22T23:52:48.9573122Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 391, in instantiated_test
2022-11-22T23:52:48.9573224Z     raise rte
2022-11-22T23:52:48.9573605Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
2022-11-22T23:52:48.9573745Z     result = test(self, **param_kwargs)
2022-11-22T23:52:48.9574094Z   File "/opt/conda/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 3472, in wrapped
2022-11-22T23:52:48.9574241Z     f(self, *args, **kwargs, coalesced=True)
2022-11-22T23:52:48.9574488Z   File "/var/lib/jenkins/workspace/test/test_sparse.py", line 1135, in test_index_select_exhaustive_index_large
2022-11-22T23:52:48.9574702Z     self._test_index_select_exhaustive_index((100, 50, 3, 3), (2, 3), device, dtype, coalesced)
2022-11-22T23:52:48.9574925Z   File "/var/lib/jenkins/workspace/test/test_sparse.py", line 1101, in _test_index_select_exhaustive_index
2022-11-22T23:52:48.9575116Z     t_sparse = t.to_sparse().coalesce() if coalesced else t.to_sparse()
2022-11-22T23:52:48.9575684Z torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 20.00 MiB (GPU 0; 7.44 GiB total capacity; 355.50 KiB already allocated; 320.00 KiB free; 2.00 MiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

I couldn't reproduce the OOM. Based on the error message, it could be CUDA context from concurrent CI tests ate up CUDA memory? cc @huydhn 


2. `pull / linux-focal-py3.7-clang7-asan / test `: https://github.com/pytorch/pytorch/actions/runs/3527607142/jobs/5917038510

This is irrelevant and already on master. See https://github.com/pytorch/pytorch/actions/runs/3518138536/jobs/5896929592

```
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr_base.h:911:14: runtime error: member call on address 0x60b000b767b0 which does not point to an object of type 'std::_Sp_counted_base<__gnu_cxx::_S_atomic>'
0x60b000b767b0: note: object has a possibly invalid vptr: abs(offset to top) too big
 dd 0f 00 2a  28 1a 80 3f f1 7f 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
              possibly invalid vptr
    #0 0x7ff139848219 in std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count(std::__weak_count<(__gnu_cxx::_Lock_policy)2> const&, std::nothrow_t) (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0xdf2e219)
    #1 0x7ff14e7972f8 in libkineto::(anonymous namespace)::configFactories() (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0x22e7d2f8)
    #2 0x7ff14e797f79 in libkineto::Config::Config() (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0x22e7df79)
    #3 0x7ff14e7b129d in libkineto::ConfigLoader::updateConfigThread() (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0x22e9729d)
    #4 0x7ff16eb2ade3  (/lib/x86_64-linux-gnu/libstdc++.so.6+0xd6de3)
    #5 0x7ff16ee32608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608)
    #6 0x7ff16ed57132 in clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/shared_ptr_base.h:911:14 in
FINISHED PRINTING LOG FILE of test_jit (/var/lib/jenkins/workspace/test/test-reports/test_jit_ansb04hr)

test_jit failed!
Traceback (most recent call last):
  File "test/run_test.py", line 1339, in <module>
    main()
  File "test/run_test.py", line 1303, in main
    raise RuntimeError("\n".join(failure_messages))
RuntimeError: test_jit failed!
```

**Comment by mrshenli** [comment]
One more failure, on `win-vs2019-cuda11.7-py3 / test (default, 2, 2, windows.8xlarge.nvidia.gpu)`: https://github.com/pytorch/pytorch/actions/runs/3527608027/jobs/5917374444

```
======================================================================
FAIL [0.003s]: test_ref_small_input_prod_cuda_int8 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 378, in instantiated_test
    result = test(self, **param_kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 828, in test_wrapper
    return test(*args, **kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\test\test_reductions.py", line 378, in test_ref_small_input
    self._test_ref(op, t)
  File "C:\actions-runner\_work\pytorch\pytorch\test\test_reductions.py", line 365, in _test_ref
    self.assertEqual(result, expected, exact_dtype=False)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 2561, in assertEqual
    assert_equal(
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_comparison.py", line 1118, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not equal!

Absolute difference: 68719476736
Relative difference: inf
```

Checked the test source code, it is op-only test, and doesn't touch module hooks at all. Also I couldn't reproduce this locally. 

https://github.com/pytorch/pytorch/blob/1cfd3858ac54fe3883534309081631a0a892ba3f/test/test_reductions.py#L373-L380

**Comment by mrshenli** [comment]
4th failure: `cuda11.6-py3.10-gcc7-sm86 / test` https://github.com/pytorch/pytorch/actions/runs/3527607833/jobs/5918576014

This is also irrelevant and on master. See https://github.com/pytorch/pytorch/actions/runs/3528028962/jobs/5917920193


```
2022-11-23T02:10:11.0209152Z Generating XML reports...
2022-11-23T02:10:11.0209491Z Generated XML report: test-reports/python-unittest/test_jit_cuda_fuser/TEST-TestCudaFuser-20221123015624.xml
2022-11-23T02:10:11.0209824Z Generated XML report: test-reports/python-unittest/test_jit_cuda_fuser/TEST-TestEnableDisableCudaFuser-20221123015624.xml
2022-11-23T02:10:11.0210162Z Generated XML report: test-reports/python-unittest/test_jit_cuda_fuser/TEST-jit.test_fuser_common.TestFuserCommon-20221123015624.xml
2022-11-23T02:10:11.0210472Z Generated XML report: test-reports/python-unittest/test_jit_cuda_fuser/TEST-TestCudaFuserOpInfoCUDA-20221123015624.xml
2022-11-23T02:10:11.0210766Z FINISHED PRINTING LOG FILE of test_jit_cuda_fuser (/var/lib/jenkins/workspace/test/test-reports/test_jit_cuda_fuser_gemnwwrr)
2022-11-23T02:10:11.0210772Z 
2022-11-23T02:10:11.0210877Z Traceback (most recent call last):
2022-11-23T02:10:11.0211031Z   File "/var/lib/jenkins/workspace/test/run_test.py", line 1339, in <module>
2022-11-23T02:10:11.0211105Z     main()
2022-11-23T02:10:11.0211255Z   File "/var/lib/jenkins/workspace/test/run_test.py", line 1314, in main
2022-11-23T02:10:11.0211359Z     raise RuntimeError(err_message)
2022-11-23T02:10:11.0211504Z RuntimeError: test_jit_cuda_fuser failed! Received signal: SIGIOT
```

**Comment by mrshenli** [comment]
One more, same as the 3rd one: `win-vs2019-cuda11.6-py3 / test`, https://github.com/pytorch/pytorch/actions/runs/3527607833/jobs/5918576620#logs

```
======================================================================
FAIL [0.003s]: test_ref_small_input_prod_cuda_int8 (__main__.TestReductionsCUDA)
Compares op against reference for small input tensors
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 378, in instantiated_test
    result = test(self, **param_kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 828, in test_wrapper
    return test(*args, **kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\test\test_reductions.py", line 378, in test_ref_small_input
    self._test_ref(op, t)
  File "C:\actions-runner\_work\pytorch\pytorch\test\test_reductions.py", line 365, in _test_ref
    self.assertEqual(result, expected, exact_dtype=False)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 2561, in assertEqual
    assert_equal(
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_comparison.py", line 1118, in assert_equal
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not equal!

Absolute difference: 68719476736
Relative difference: inf
```

**Pull request closed without merging**
